src/ucp/tag/offload.h

```diff
@@ -87,4 +90,55 @@ ucp_tag_offload_try_cancel(ucp_worker_t *worker, ucp_request_t *req, int force)
     }
 }
 
+/* For homogeneous clusters it is expected that tag offload messages will arrive
+ * from the single interface. Otherwise need to add a particular iface to tags
+ * hash, where key is a tag masked with tag_sender mask. */
+static UCS_F_ALWAYS_INLINE void
+ucp_tag_offload_unexp(ucp_worker_iface_t *iface, ucp_tag_t tag)
+{
+    ucp_worker_t *worker   = iface->worker;
+    ucp_context_t *context = worker->context;
+    ucp_tag_t tag_key;
+    khiter_t hash_it;
+    int ret;
+
+    ++iface->proxy_recv_count;
+
+    if (ucs_unlikely(!(iface->flags & UCP_WORKER_IFACE_FLAG_ON_OFFLOAD_QUEUE))) {
+
+        if (ucs_unlikely(!context->config.ext.tm_offload)) {
+            /* Can happen when different processes have different configuration,
+             * which is very uncommon. */
+            return;
+        }
+
+        if (ucs_queue_is_empty(&worker->tm.offload.ifaces)) {
+            ucs_assert(worker->tm.offload.thresh       == SIZE_MAX);
+            ucs_assert(worker->tm.offload.zcopy_thresh == SIZE_MAX);
+
+            worker->tm.offload.thresh       = context->config.ext.tm_thresh;
+            worker->tm.offload.zcopy_thresh = context->config.ext.tm_max_bcopy;
+
+            ucs_debug("Enable TM offload: thresh %zu, zcopy_thresh %zu",
+                    worker->tm.offload.thresh, worker->tm.offload.zcopy_thresh);
+        }
+
+        ucs_queue_push(&worker->tm.offload.ifaces, &iface->queue);
```

Comment:
is there any point of keeping a queue of offload ifaces now? maybe just a single pointer in case there is single iface, and NULL if there are 0 or more than one?

Reply:
Is there any scenario when we can get back to single interface from several ones? Seems like no..

Reply:
no, we never deactivate..

Reply:
So will make a union of {hash/iface pointer} at this point